### PR TITLE
Fix downloader hangs in some case

### PR DESCRIPTION
**Pull request by asdacap**
Fix downloader hangs when peer response with no body.

## Changes:
- Allow downloader and merge block downloader to process block partially.
- In RequestBodies, exit early if peer response with no body at all.
- In RequestReceipts, exit early if block does not have body when it should have body.
- In DownloadBodies, throw exception when block should have body but does not.

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [X] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [X] Yes
- [ ] No

**Comments about testing , should you have some** (optional)

## Further comments (optional)

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/Nethermind/Nethermind.Synchronization/Blocks/BlockDownloader.cs [hunk 1]
```diff
@@ -475,7 +486,13 @@ protected async Task RequestReceipts(PeerInfo peer, CancellationToken cancellati
                 for (int i = 0; i < result.Length; i++)
                 {
                     TxReceipt[] txReceipts = result[i];
-                    if (!context.TrySetReceipts(i + offset, txReceipts, out Block block))
+                    Block block = context.GetBlockByRequestIdx(i + offset);
+                    if (block.Header.HasBody && block.Transactions.Length == 0)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/Blocks/BlockDownloader.cs:490]: what about unlces only blocks?

Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/Blocks/BlockDownloader.cs:490]: Yup, good catch

File: src/Nethermind/Nethermind.Merge.Plugin/Synchronization/MergeBlockDownloader.cs [hunk 2]
```diff
@@ -172,6 +172,11 @@ bool HasMoreToSync()
                     Block currentBlock = blocks[blockIndex];
                     if (_logger.IsTrace) _logger.Trace($"Received {currentBlock} from {bestPeer}");
 
+                    if (currentBlock.Header.HasBody && currentBlock.Body.IsEmpty)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/Synchronization/MergeBlockDownloader.cs:175]: This is now duplicated in 3 places, should we have some kind of method for this validation?

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Merge.Plugin/Synchronization/MergeBlockDownloader.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Merge.Plugin/Synchronization/MergeBlockDownloader.cs:175]: Probably, I'm still good with this expression IMO, still understandable.

Comment (asdacap) [reply src/Nethermind/Nethermind.Merge.Plugin/Synchronization/MergeBlockDownloader.cs]: Renamed to `currentBlock.IsBodyMissing`

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**